kong/plugins/redirect/handler.lua

```diff
@@ -0,0 +1,73 @@
+local kong = kong
+local kong_meta = require "kong.meta"
+local socket_url = require "socket.url"
+
+local RedirectHandler = {}
+
+-- Priority 779 so that it runs after all rate limiting/validation plugins
+-- and all transformation plugins, but before any AI plugins which call upstream
+RedirectHandler.PRIORITY = 779
+RedirectHandler.VERSION = kong_meta.version
+
+function RedirectHandler:access(conf)
+    -- Use the 'location' as-is as the default
+    -- This is equivalent to conf.incoming_path == 'ignore'
+    local location = conf.location
+
+    if conf.incoming_path ~= "ignore" then
+        -- Parse the URL in 'conf.location' and the incoming request
+        request_path = kong.request.get_path_with_query()
+        local request_path_url = socket_url.parse(request_path)
+        local location_path_url = socket_url.parse(location)
+
+        -- The path + query are different depending on the 'incoming_path' configuration
+        local path = ""
+        local query = ""
+
+        -- If incoming_path == 'keep', use the incoming request query
+        if conf.incoming_path == "keep" then
+            query = request_path_url.query
+            path = request_path_url.path;
+        end
+
+        -- If it's 'merge', merge the incoming request path+query with the location path+query
+        if conf.incoming_path == "merge" then
```

Comment:
I'm open to removing the `merge` behaviour entirely until someone asks for it. I added the logic as something that may be useful, not something that's been specifically requested.